src/extremeweatherbench/utils.py

```diff
@@ -403,3 +320,176 @@ def read_event_yaml(input_pth: str | Path) -> dict:
+        forecast_ds["latitude"].min().values,
+        forecast_ds["latitude"].max().values,
+        forecast_ds["longitude"].min().values,
+        forecast_ds["longitude"].max().values,
+    )
+    # Uses indexing in the dataframe to capture metadata columns for future use
+    point_obs_metadata = case_subset_point_obs_df[point_obs_metadata_vars]
+
+    # Reset index to allow for easier modification
+    case_subset_point_obs_df = case_subset_point_obs_df.reset_index()
+    case_subset_point_obs_df.loc[:, "station"] = case_subset_point_obs_df[
+        "station"
+    ].astype("str")
+
+    # Set up multiindex to enable slicing along individual timesteps
+    case_subset_point_obs_df = case_subset_point_obs_df.set_index(
+        ["station", "time"]
+    ).sort_index()
+
+    aligned_forecast_list = []
+    aligned_observation_list = []
+
+    # Loop init and lead times to prevent indexing error from duplicate valid times
+    for init_time, lead_time in itertools.product(
+        forecast_ds.init_time, forecast_ds.lead_time
+    ):
+        valid_time = pd.Timestamp(
+            init_time.values + pd.to_timedelta(lead_time.values, unit="h").to_numpy()
+        )
+        valid_time_index = pd.IndexSlice[:, valid_time]
+        obs_timeslice = (
+            case_subset_point_obs_df.loc[valid_time_index, :]
+            if valid_time in case_subset_point_obs_df.index.get_level_values(1)
+            else None
+        )
+        if obs_timeslice is None:
+            continue
+        obs_timeslice = obs_timeslice.reset_index()
+
+        station_ids = obs_timeslice["station"]
+        lons = xr.DataArray(obs_timeslice["longitude"].values, dims="station")
+        lats = xr.DataArray(obs_timeslice["latitude"].values, dims="station")
+
+        forecast_at_obs_ds = forecast_ds.sel(
+            init_time=init_time, lead_time=lead_time
+        ).interp(latitude=lats, longitude=lons, method="nearest")
+        forecast_at_obs_ds = forecast_at_obs_ds.assign_coords(
+            {"station": station_ids, "time": valid_time}
+        )
+        forecast_at_obs_ds.coords["lead_time"] = lead_time
+        forecast_at_obs_ds.coords["init_time"] = init_time
+        valid_time_subset_obs_ds = obs_timeslice.to_xarray().set_coords(
+            point_obs_metadata
+        )[data_var]
+        valid_time_subset_obs_ds = valid_time_subset_obs_ds.swap_dims(
+            {"index": "station"}
+        )
+        valid_time_subset_obs_ds.coords["lead_time"] = lead_time
+        valid_time_subset_obs_ds.coords["init_time"] = init_time
+
+        aligned_observation_list.append(valid_time_subset_obs_ds)
+        aligned_forecast_list.append(forecast_at_obs_ds)
+    # concat the dataarrays along the station dimension
+    interpolated_forecast = xr.concat(aligned_forecast_list, dim="station")
+    interpolated_observation = xr.concat(aligned_observation_list, dim="station")
+    return (interpolated_forecast, interpolated_observation)
```

Comment:
I think it's okay to keep these outputs like this, because they're logically named and it's easier to read this way.